### PR TITLE
[SPARK-45283][CORE][TESTS] Make StatusTrackerSuite less fragile

### PR DESCRIPTION
**Pull request by xiongbo-sjtu**
### Why are the changes needed?

It's discovered from [Github Actions](https://github.com/xiongbo-sjtu/spark/actions/runs/6270601155/job/17028788767) that StatusTrackerSuite can run into random failures, as shown by the following error message.  The proposed fix is to update the unit test to remove the nondeterministic behavior.

```
    [info] StatusTrackerSuite:
    [info] - basic status API usage (99 milliseconds)
    [info] - getJobIdsForGroup() (56 milliseconds)
    [info] - getJobIdsForGroup() with takeAsync() (48 milliseconds)
    [info] - getJobIdsForGroup() with takeAsync() across multiple partitions (58 milliseconds)
    [info] - getJobIdsForTag() *** FAILED *** (10 seconds, 77 milliseconds)
    [info] The code passed to eventually never returned normally.
           Attempted 651 times over 10.005059944000001 seconds. 
           Last failure message: Set(3, 2, 1) was not equal to Set(1, 2). (StatusTrackerSuite.scala:148)
```

Full trace can be found [here](https://issues.apache.org/jira/browse/SPARK-45283).

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
```
build/mvn package -DskipTests -pl core
build/mvn -Dtest=none -DwildcardSuites=org.apache.spark.StatusTrackerSuite test
```

### Was this patch authored or co-authored using generative AI tooling?
No

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
Thank you for making a PR, @xiongbo-sjtu . 

It seems that something is missing in your PR description. Where is the `red` part? Maybe, you didn't copy it from Apache Spark JIRA to here?
> because FutureAction.jobIds is not a sorted sequence (by design), as shown in the following stack trace (highlighted in red).

**Comment by LuciferYang** [comment]
> as shown in the following stack trace (highlighted in red)

which stack? Seems some information is missing in the PR description?

File: core/src/test/scala/org/apache/spark/StatusTrackerSuite.scala [hunk 1]
```diff
@@ -33,7 +33,7 @@ class StatusTrackerSuite extends SparkFunSuite with Matchers with LocalSparkCont
     val jobId: Int = eventually(timeout(10.seconds)) {
       val jobIds = jobFuture.jobIds
       jobIds.size should be(1)
-      jobIds.head
+      jobIds.sorted.reverse.head
```
Comment (LuciferYang) [inline core/src/test/scala/org/apache/spark/StatusTrackerSuite.scala:36]: If the goal is to get the largest JobId, can we use `jobIds.max`?

Comment (xiongbo-sjtu) [reply core/src/test/scala/org/apache/spark/StatusTrackerSuite.scala]: Good point.  Updated the code in this PR.

**Comment by xiongbo-sjtu** [comment]
@dongjoon-hyun I've updated the PR description.

File: core/src/test/scala/org/apache/spark/StatusTrackerSuite.scala [hunk 2]
```diff
@@ -33,7 +33,7 @@ class StatusTrackerSuite extends SparkFunSuite with Matchers with LocalSparkCont
     val jobId: Int = eventually(timeout(10.seconds)) {
       val jobIds = jobFuture.jobIds
       jobIds.size should be(1)
-      jobIds.head
+      jobIds.max
```
Comment (mridulm) [inline core/src/test/scala/org/apache/spark/StatusTrackerSuite.scala:36]: Revert ? `jobIds.size` is 1, so `head` == `max`

File: core/src/test/scala/org/apache/spark/StatusTrackerSuite.scala [hunk 3]
```diff
@@ -61,7 +61,7 @@ class StatusTrackerSuite extends SparkFunSuite with Matchers with LocalSparkCont
     // Passing `null` should return jobs that were not run in a job group:
     val defaultJobGroupFuture = sc.parallelize(1 to 1000).countAsync()
     val defaultJobGroupJobId = eventually(timeout(10.seconds)) {
-      defaultJobGroupFuture.jobIds.head
+      defaultJobGroupFuture.jobIds.max
     }
```
Comment (mridulm) [inline core/src/test/scala/org/apache/spark/StatusTrackerSuite.scala:65]: Revert ? There should only be a single job submitted here, so `head` will be `max`

File: core/src/test/scala/org/apache/spark/StatusTrackerSuite.scala [hunk 4]
```diff
@@ -92,7 +92,7 @@ class StatusTrackerSuite extends SparkFunSuite with Matchers with LocalSparkCont
     sc.statusTracker.getJobIdsForGroup("my-job-group2") shouldBe empty
     val firstJobFuture = sc.parallelize(1 to 1000, 1).takeAsync(1)
     val firstJobId = eventually(timeout(10.seconds)) {
-      firstJobFuture.jobIds.head
+      firstJobFuture.jobIds.max
     }
```
Comment (mridulm) [inline core/src/test/scala/org/apache/spark/StatusTrackerSuite.scala:96]: We can revert this as well - same as above.

File: core/src/test/scala/org/apache/spark/StatusTrackerSuite.scala [hunk 5]
```diff
@@ -105,7 +105,7 @@ class StatusTrackerSuite extends SparkFunSuite with Matchers with LocalSparkCont
     sc.statusTracker.getJobIdsForGroup("my-job-group2") shouldBe empty
     val firstJobFuture = sc.parallelize(1 to 1000, 2).takeAsync(999)
     eventually(timeout(10.seconds)) {
-      firstJobFuture.jobIds.head
+      firstJobFuture.jobIds.max
     }
```
Comment (mridulm) [inline core/src/test/scala/org/apache/spark/StatusTrackerSuite.scala:109]: Revert, same as above.

File: core/src/test/scala/org/apache/spark/StatusTrackerSuite.scala [hunk 6]
```diff
@@ -121,7 +121,7 @@ class StatusTrackerSuite extends SparkFunSuite with Matchers with LocalSparkCont
     // countAsync()
     val firstJobFuture = sc.parallelize(1 to 1000).countAsync()
     val firstJobId = eventually(timeout(10.seconds)) {
-      firstJobFuture.jobIds.head
+      firstJobFuture.jobIds.max
     }
```
Comment (mridulm) [inline core/src/test/scala/org/apache/spark/StatusTrackerSuite.scala:125]: Revert - same as above.

**Review by mridulm — COMMENTED** [review]
Thanks for sending this PR !
Let us use the minimal subset of change required.

**Comment by mridulm** [comment]
Looking into this more, I dont think this PR will fix the issue being seen.
The exception indicates this is not a sorting issue - it is two `Set's` being compared.

The number of entries is incorrect - 2 jobs were expected for each of the two tags, but 3 were found.
We will have to find out what the race condition here is.

**Comment by mridulm** [comment]
So looking at it more, the race condition is due to the following:

* We have two jobs for the `takeAsync(999)` - but they are executed one after another.
* In most cases, the ` val thirdJobId = eventually(timeout(10.seconds)) {` will detect the first job start and return that value.
* Subsequently, the last eventually starts and it will immediately match `sc.statusTracker.getJobIdsForTag("tag2").toSet should be (Set(secondJobId, thirdJobId))` - since the second job for `takeAsync` has not yet started.

In some corner cases, the second job for `takeAsync` would have started before the last `eventually` - and so now we have 3 jobs tagged with `tag2` instead of 2.

To fix this, IMO we should do the following:

```
  val thirdJobIds = eventually(timeout(10.seconds)) {
      if (thirdJobFuture.jobIds.size != 2) throw new IllegalStateException("Both jobs not yet started")
      thirdJobFuture.jobIds
    }

 eventually(timeout(10.seconds)) {
      sc.statusTracker.getJobIdsForTag("tag1").toSet should be (
        Set(firstJobId, secondJobId))
      sc.statusTracker.getJobIdsForTag("tag2").toSet should be (
        Set(secondJobId) ++ , thirdJobIds)
    }
```

(Paraphrasing - some such).

**Comment by xiongbo-sjtu** [comment]
I've updated the PR to ensure that the following conditions are met, before `thirdJobFuture` is even constructed.
- [Condition A] firstJobFuture.isCompleted should be(true)
- [Condition B] secondJobFuture.isCompleted should be(true)
- [Condition C] sc.getJobTags.size should be(1)
- [Condition D] sc.statusTracker.getJobIdsForTag("tag1").toSet.size should be(2)

Conditions A & B tell us that the first 2 future actions are completed (i.e., no more state transition).
Condition C means that only 1 tag remains (i.e., `tag1` is removed and `tag2` is kept).
Condition D means that only 2 jobs have been marked with `tag1`.

Given that, we'd expect that the reported issue disappears in the sense that `sc.statusTracker.getJobIdsForTag("tag1").toSet` won't return a set of 3 items.

File: core/src/test/scala/org/apache/spark/StatusTrackerSuite.scala [hunk 7]
```diff
@@ -140,6 +140,14 @@ class StatusTrackerSuite extends SparkFunSuite with Matchers with LocalSparkCont
...
 
     sc.removeJobTag("tag1")
+    eventually(timeout(60.seconds)) {
+      // Make sure that only two jobs are tagged after tag1
+      firstJobFuture.isCompleted should be(true)
+      secondJobFuture.isCompleted should be(true)
+      sc.getJobTags should be (Set("tag2"))
+      sc.statusTracker.getJobIdsForTag("tag1").toSet.size should be(2)
+    }
+
```
Comment (mridulm) [inline core/src/test/scala/org/apache/spark/StatusTrackerSuite.scala:150]: This is not going to fix the issue - I have detailed the fix required for address the flakiness.

**Comment by mridulm** [comment]
@xiongbo-sjtu to clarify, when I mentioned about two jobs, I was not referring to `firstJobFuture` and `secondJobFuture` - but two jobs submitted as part of `takeAsync's` implementation when `thirdJobFuture` executes.

**Comment by xiongbo-sjtu** [comment]
> when I mentioned about two jobs, I was not referring to `firstJobFuture` and `secondJobFuture` - but two jobs submitted as part of `takeAsync's` implementation when `thirdJobFuture` executes.

Make sense.  Updated the PR.

**Comment by mridulm** [comment]
Merged to master.
Thanks for fixing this @xiongbo-sjtu !
Thanks for the review @dongjoon-hyun, @LuciferYang :-)

**Comment by LuciferYang** [comment]
@xiongbo-sjtu Would it be convenient to submit a similar patch for branch-3.5? I found that this case would also likely fail in branch-3.5.

https://github.com/apache/spark/actions/runs/6510937376/job/17685699603

**Comment by xiongbo-sjtu** [comment]
> @xiongbo-sjtu Would it be convenient to submit a similar patch for branch-3.5? I found that this case would also likely fail in branch-3.5.
> 
> https://github.com/apache/spark/actions/runs/6510937376/job/17685699603

Here we go.

https://github.com/apache/spark/pull/43388

**Pull request closed without merging**